### PR TITLE
Major update to nlmixr2 conversion

### PR DESCRIPTION
**Pull request by JohanWBorg**
Updated script structure along with major updates to, but not limited to
- More adaptive error model handling and multiple add/prop terms
- Handling of fixed effects and "SAME" function in NONMEM
- Handling of extra expression in DES block
- Use NONMEM estimated ETAs when validating converted model

### CODE REVIEW

File: src/pharmpy/plugins/nlmixr/CodeGenerator.py
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/CodeGenerator.py:1]: Since the file names are also module names we shouldn't use caps in filenames. use `code_generator.py` for example. This module might also be of value outside of this plugin so it could be moved to pharmpy/internals

File: src/pharmpy/plugins/nlmixr/create_ini.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    Add THETAs to code generator when converting a model
+
+    Parameters
+    ----------
+    model : pharmpy.model.Model
+        Pharmpy model object.
+    cg : CodeGenerator
+        Code generator to add code upon.
+
+    Returns
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/create_ini.py:20]: Remove Returns since it doesn't return anything.

File: src/pharmpy/plugins/nlmixr/create_ini.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    Add ETAs to code generator when converting a model
+
+    Parameters
+    ----------
+    model : pharmpy.model.Model
+        Pharmpy model object.
+    cg : CodeGenerator
+        Code generator to add code upon.
+
+    Returns
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/create_ini.py:45]: Remove Returns

File: src/pharmpy/plugins/nlmixr/create_ini.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    Add SIGMAs to code generator when converting a model.
+
+    Parameters
+    ----------
+    model : pharmpy.model.Model
+        Pharmpy model object.
+    cg : CodeGenerator
+        Code generator to add code upon.
+
+    Returns
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/create_ini.py:77]: Remove returns

File: src/pharmpy/plugins/nlmixr/error_model.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+    Raises
+    ------
+    ValueError
+        If the model either has multiple additative- or proportional error
+        terms, the function will raise a ValueError
+
+    Returns
+    -------
+    res : Union[sympy.Symbol, sympy.Add]
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/error_model.py:33]: I am not sure this is the correct way to document returning a tuple, but I might be wrong.

File: src/pharmpy/plugins/nlmixr/error_model.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+    ValueError
+        Will raise ValueError if model has defined error model that does not
+        match the format of the found error terms.
+
+    Returns
+    -------
+    None
+        Modifies the given CodeGenerator object. Returns nothing
+
+    Example
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/error_model.py:134]: Remove empty example

File: src/pharmpy/plugins/nlmixr/error_model.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    Parameters
+    ----------
+    cg : CodeGenerator
+        Codegenerator object holding the code to be added to.
+    error : dict
+        Dictionary with additive and proportional error terms.
+    symbol : str
+        Symbol of dependent variable.
+
+    Returns
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/error_model.py:181]: Remove empty Returns

File: src/pharmpy/plugins/nlmixr/model.py [hunk 7]
```diff
@@ -4,52 +4,60 @@
...
+    keep_etas : bool, optional
+        Decide if NONMEM estimated thetas are to be used. The default is False.
+    skip_check : bool, optional
+        Skip determination of error model type. Could speed up conversion. The default is False.
 
-    def add(self, line):
-        self.lines.append(f'{" " * self.indent_level}{line}')
+    Returns
+    -------
+    pharmpy.model
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/model.py:51]: Should it be pharmpy.model.Model?

File: src/pharmpy/plugins/nlmixr/model.py [hunk 8]
```diff
@@ -62,15 +70,35 @@ def convert_model(model):
...
+        # FIXME: dropping columns runs update source which becomes redundant.
+        # drop_dropped_columns(nlmixr_model)
+        if all(x in nlmixr_model.dataset.columns for x in ["RATE", "DUR"]):
+            nlmixr_model = drop_columns(nlmixr_model, ["DUR"])
+        nlmixr_model = nlmixr_model.replace(
+            datainfo=nlmixr_model.datainfo.replace(path=None),
+            dataset=nlmixr_model.dataset.reset_index(drop=True),
+        )
+    else:
+        print_warning("No connected dataset or dataset is empty")
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/model.py:91]: No need for warning here. This is a legitimate case

File: src/pharmpy/plugins/nlmixr/model.py [hunk 9]
```diff
@@ -62,15 +70,35 @@ def convert_model(model):
...
+    nlmixr_model = check_model(nlmixr_model, skip_error_model_check=skip_check)
 
-def name_mangle(s):
-    return s.replace('(', '').replace(')', '').replace(',', '_')
+    nlmixr_model.update_source()
+
+    return nlmixr_model
 
 
 class ExpressionPrinter(sympy_printing.str.StrPrinter):
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/model.py:104]: Wasn't the `ExpressionPrinter` moved to another module or are there two printers now?

File: src/pharmpy/plugins/nlmixr/model.py [hunk 10]
```diff
@@ -93,96 +121,166 @@ def _print_Function(self, expr):
...
+    if "EVID" not in temp_model.dataset.columns:
+        temp_model.dataset["EVID"] = get_evid(temp_model)
+    return temp_model
 
 
 @dataclass
 class NLMIXRModelInternals:
     src: Optional[str] = None
     path: Optional[Path] = None
+    nonmem_control_stream: Optional = None
```
Comment (rikardn) [inline src/pharmpy/plugins/nlmixr/model.py:283]: Why is this needed? Shouldn't everything about NONMEM be forgotten after conversion?

**Review by rikardn — CHANGES_REQUESTED** [review]
Nice work! I have some comments

File: src/pharmpy/plugins/nlmixr/CodeGenerator.py
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/CodeGenerator.py:1]: Moved file to pharmpy/internals and changed the name to `code_genarator.py`

File: src/pharmpy/plugins/nlmixr/create_ini.py
[the remark this answers is not in the document]
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_ini.py:1]: Corrected the file name to `ini.py` instead

File: src/pharmpy/plugins/nlmixr/create_model_block.py
[the remark this answers is not in the document]
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_model_block.py:1]: Corrected name to `model_block.py`. (`model.py` is already in use and model_block seems a bit more descriptive)

File: src/pharmpy/plugins/nlmixr/create_ini.py (hunk 1 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_ini.py:20]: Removed

File: src/pharmpy/plugins/nlmixr/create_ini.py (hunk 2 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_ini.py:45]: Removed

File: src/pharmpy/plugins/nlmixr/create_ini.py (hunk 3 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_ini.py:77]: Removed

File: src/pharmpy/plugins/nlmixr/create_ini.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    ----------
+    cg : CodeGenerator
+        Codegenerator object holding the code to be added to.
+    parameter : sympy.Symbol
+        The parameter to be added. Either theta or sigma
+    boundary : bool, optional
+        Decide if the parameter should be added with or without parameter
+        boundries. The default is False.
+
+    Returns
```
[the remark this answers is not in the document]
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_ini.py:105]: Removed

File: src/pharmpy/plugins/nlmixr/create_model_block.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+    Parameters
+    ----------
+    model : pharmpy.model.Model
+        A pharmpy model object to add statements to.
+    cg : CodeGenerator
+        Codegenerator object holding the code to be added to.
+    statements : pharmpy.model.statements
+        Statements to be added to the code generator.
+
+    Returns
```
[the remark this answers is not in the document]
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_model_block.py:48]: Removed

File: src/pharmpy/plugins/nlmixr/create_model_block.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+    Add the ODEs from a model to a code generator
+
+    Parameters
+    ----------
+    model : pharmpy.model.Model
+        A pharmpy model object to add ODEs to.
+    cg : CodeGenerator
+        Codegenerator object holding the code to be added to.
+
+    Returns
```
[the remark this answers is not in the document]
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/create_model_block.py:132]: Removed

File: src/pharmpy/plugins/nlmixr/error_model.py (hunk 4 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/error_model.py:33]: This was the style of the automatically generated docstring, which I chose to use. If you'd like me to change it, please let me know! Otherwise, it will remain the same.

File: src/pharmpy/plugins/nlmixr/error_model.py (hunk 5 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/error_model.py:134]: Removed. Also removed unnecessary return

File: src/pharmpy/plugins/nlmixr/error_model.py (hunk 6 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/error_model.py:181]: Removed

File: src/pharmpy/plugins/nlmixr/model.py (hunk 7 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/model.py:51]: Yes, my mistake, this has been corrected to `pharmpy.model.Model`

File: src/pharmpy/plugins/nlmixr/model.py (hunk 8 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/model.py:91]: Removed warning

File: src/pharmpy/plugins/nlmixr/model.py (hunk 9 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/model.py:104]: Correct. This has now been removed as it should have been and is only defined in ìni.py`

File: src/pharmpy/plugins/nlmixr/model.py (hunk 10 above)
Comment (JohanWBorg) [reply src/pharmpy/plugins/nlmixr/model.py:283]: This is implemented to be able to convert statements within the `$DES` block. Since these are not saved in `model.statements`, they need to be saved in some other place for the nlmixr2 model. I decided to save the entire control stream from NONMEM. But as you stated, I agree that NONMEM should be more or less forgotten after conversion. 
Since the DES statements still need to be saved in one place or another, I modified the code to only save the DES statements instead of the entire control stream but I am happy to receive any other ideas or thoughts on other implementations

**Comment by JohanWBorg** [comment]
All comments have been fixed, if there are any issues with the implementations or any other comments, please let me know! :D

File: src/pharmpy/plugins/nlmixr/error_model.py (hunk 4 above)
Comment (rikardn) [reply src/pharmpy/plugins/nlmixr/error_model.py:33]: Ok, let's keep it. Since it is not the top level API docstring it is not super important

File: src/pharmpy/plugins/nlmixr/model.py (hunk 10 above)
Comment (rikardn) [reply src/pharmpy/plugins/nlmixr/model.py]: Ok, I understand. Since we regenerate the code from scratch this is needed as a workaround.

File: MANIFEST.in [hunk 14]
```diff
@@ -33,3 +33,12 @@ recursive-include src *.py
 
 # added by check-manifest
 recursive-include src *.py
+
+# added by check-manifest
```
Comment (rikardn) [inline MANIFEST.in:37]: These seems to have been added unnecessarily

**Comment by rikardn** [comment]
I'll merge and fix the MANIFEST thing. It sees to have been added extra times before. Good work!

**Comment by JohanWBorg** [comment]
Thank you so much for your help and comments! :)
